clap_derive/src/derives/subcommand.rs

```diff
@@ -131,8 +131,34 @@ fn gen_augment(
 
             match &*kind {
                 Kind::ExternalSubcommand => {
-                    quote_spanned! { kind.span()=>
-                        let app = app.setting(clap::AppSettings::AllowExternalSubcommands);
+                    let ty = match variant.fields {
+                        Unnamed(ref fields) if fields.unnamed.len() == 1 => &fields.unnamed[0].ty,
+
+                        _ => abort!(
+                            variant,
+                            "The enum variant marked with `external_attribute` must be \
+                             a single-typed tuple, and the type must be either `Vec<String>` \
+                             or `Vec<OsString>`."
+                        ),
+                    };
+                    match subty_if_name(ty, "Vec") {
+                        Some(subty) => {
+                            if is_simple_ty(subty, "OsString") {
+                                quote_spanned! { kind.span()=>
+                                    let app = app.setting(clap::AppSettings::AllowExternalSubcommands).setting(clap::AppSettings::AllowInvalidUtf8ExternalSubcommands);
+                                }
+                            } else {
```

Comment:
Shouldn't this be `else if is_simple_ty(subty, "String")` and then have an abort in `else`?

Since that `else` and `None` after that have the same abort message, we need to restructure this to be kept DRY. To be honest, even the message at the beginning can be combined since it's same.

Reply:
> Shouldn't this be else if is_simple_ty(subty, "String") and then have an abort in else?

I deferred the handling of that case to later. since this case is completely agnostic to what types are supported.  This ensures we can update the other case without worrying about this case being updated, unless a new OsStr type is added.

> Since that else and None after that have the same abort message, we need to restructure this to be kept DRY. To be honest, even the message at the beginning can be combined since it's same.

Not convinced that sharing the message is sufficient for DRY.    DRY carries a large cost with it that needs to be weighed out with the gains.

I could see us adding an `enum SubcommandType`  and parsing into that.   I do not think that refactoring should be bundled in with this PR though.

Reply:
> This ensures we can update the other case without worrying about this case being updated

That makes sense.

Reply:
In that case, shouldn't the abort messages have different phrases?

Reply:
Could you expand on what you are referring to?

Reply:
Continuing this discussion in the other comment.